### DISCUSSION TITLE
GUI screensaver

### DISCUSSION THREAD

**Discussion by MickTheMechanic** [ImpulseAdventure/GUIslice#335] [Q&A] [answered]
Hi guys, Just looking for an Idea of how to trigger a screen saver.

I simply want to display a certain page if the screen hasn't been touched for a certain amount of time.

What is the simplest way to check if the screen has been touched? 

Mick--

**Comment by bcodermicro** [comment]
Have every touch reset a tiimer? If the timer gets to a certaiin value then initiate the "screen savver"

**Reply by MickTheMechanic** [reply]
Thats exactly what I plan to do, the question is whats the best way to register a touch to reset the timer? As in, is there some kind of function within GUISlice that I can use?

**Comment by bcodermicro** [comment, marked as answer]
There is a callback on every touch element you set up. Inside of that callback reset the time along with ddoing whatever the button should do.

**Reply by MickTheMechanic** [reply]
I had also thought about that, but would I then need to add code to every touch element? Or is there somewhere I could place the code so that it is triggered by all touch elements?

**Reply by MickTheMechanic** [reply]
Its working, thanks alot for your help.

**Reply by bcodermicro** [reply]
wrrite a function for the time reset!

**Reply by MickTheMechanic** [reply]
its working, I have already written the code. I was just unsure how to trigger the reset, but I simply used if ( eTouch == GSLC_TOUCH_UP_IN ) {    to trigger the reset.

**Reply by jjriggs** [reply]
Mick, how did you get this working? I'm working on a similar problem to turn a backlight on or off after a timeout or a touch, but I can't figure out how to register a touch outside a GUI element. I tried doing it like this:


```
// This part works
touch_event = gslc_GetTouch(&m_gui, &nTouchX, &nTouchY, &nTouchZ, &eInputEvent, &nInputVal);
if(touch_event)
{
  //backlight on and reset timer
}
else
{
  //turn backlight off if timer expired
}

// But then this does not
gslc_Update(&m_gui);
```
The problem is that calling `GetTouch` clears out the touch event queue, which breaks the `Update` call afterward.

**Reply by MickTheMechanic** [reply]
I have it set up as following.. 
```

unsigned long screensaverTimer;
unsigned long currentMillis;
const uint16_t Minutes = 60000;
uint8_t lastPageId = 2;


if ( eTouch == GSLC_TOUCH_UP_IN ) {
    screensaverTimer = currentMillis;
    // From the element's ID we can determine which button was pressed.
    switch (pElem->nId) {
//<Button Enums !Start!>
case WAKE_BTN:
        gslc_SetPageCur(&m_gui, lastPageId);
        break;

//........................//

void setup (){

screensaverTimer = millis();

//........................//

void loop (){

currentMillis = millis();

 if (currentMillis - screensaverTimer >= 1 * Minutes) {
      lastPageId = gslc_GetPageCur(&m_gui) ;
      gslc_SetPageCur(&m_gui, PG_ScreenSaver);
    }
}

```

Wake button is just a box with touch enabled that covers the whole screensaver.
Any time any button is pressed the screensaver timer resets.
Before the screensaver page is opened the current page id is saved, this allows the device to "wake" back to its previous state.

**Reply by jjriggs** [reply]
Hmm, that's an interesting solution. I don't think that will work for me though, because I want to keep displaying the same page while the backlight is dimmed. Putting an invisible "wake" button over the UI would block all other input events.
Thanks for the feedback though!

**Reply by ImpulseAdventure** [reply]
The screen-sized button is an interesting workaround to provide for a screensaver wake-up. If one changed the “fill enable” on this button to false, one could still redraw the underlying elements but the touches would not be registered unless this button were marked disabled after the wake up.

Alternately, I am thinking it might be worth me adding a simple TouchAny callback that can be optionally enabled / disabled. Any touch event (even if it doesn’t register with an element) would invoke the callback. This could provide an easy way to facilitate a touch-based wake up action. No need to create extra buttons.

Thoughts?

**Reply by jjriggs** [reply]
Impulse, a "TouchAny" callback or something like it is exactly the sort of thing I was looking for.

**Reply by jjriggs** [reply]
The other solution I was thinking about is a variation of `GetTouch` that does not flush the event queue. In other words, it just returns the latest event, but doesn't clear it out, so that subsequent calls to `Update` still get that event.

**Reply by MickTheMechanic** [reply]
What would happen if you put a box behind the other elements instead of in front of them? Might be a temporary solution for you.
